toolchain/lex/lex.cpp

```diff
@@ -93,19 +97,61 @@ class [[clang::internal_linkage]] Lexer {
 
   auto current_line() -> LineIndex { return LineIndex(line_index_); }
 
-  auto current_line_info() -> TokenizedBuffer::LineInfo* {
+  auto current_line_info() -> LineInfo* {
     return &buffer_.line_infos_[line_index_];
   }
 
-  auto ComputeColumn(ssize_t position) -> int {
-    CARBON_DCHECK(position >= current_line_info()->start);
-    return position - current_line_info()->start;
-  }
+  auto next_line() -> LineIndex { return LineIndex(line_index_ + 1); }
 
-  auto NoteWhitespace() -> void {
-    buffer_.token_infos_.back().has_trailing_space = true;
+  auto next_line_info() -> LineInfo* {
+    CARBON_DCHECK(line_index_ + 1 <
+                  static_cast<ssize_t>(buffer_.line_infos_.size()));
+    return &buffer_.line_infos_[line_index_ + 1];
   }
 
+  auto NoteWhitespace() -> void;
+
+  auto AddLexedToken(TokenInfo info) -> TokenIndex;
+
+  // A collection of overloads that provide type-checked APIs for lexing
+  // specific kinds of tokens and providing the appropriate payload data for
+  // each of them.
+  //
+  // The base overload allows a runtime token kind for any kind that doesn't
+  // require a payload at all. Token kinds that require a payload are an error
+  // to use with this overload.
+  auto LexToken(TokenKind kind, int32_t byte_offset) -> TokenIndex;
```

Comment:
I think the first overload should be separated from the rest, because it has distinctly different semantics (e.g. the meaning of the first parameter). Among other things, I think we can document the remaining overloads more clearly if we're not documenting the first one at the same time:

```suggestion
  // Lexes a token with the given kind at the given offset. `kind` must not be
  // a kind that requires a payload.
  auto LexToken(TokenKind kind, int32_t byte_offset) -> TokenIndex;

  // Lexes a token with the given kind at the given offset, with the given payload.
  // `Payload` must be the payload type associated with `Kind`.
  template <TokenKind::RawEnumType Kind, typename Payload>
  auto LexTokenWithPayload(Payload payload, int32_t byte_offset) = delete;
```

If you make this change, I think you can probably drop the comments from the overloads below. It might also be worthwhile to express the overloads as specializations, e.g:

```cpp
  template <>
  auto LexToken<TokenKind::Identifier>(IdentifierId id, int32_t byte_offset) -> TokenIndex;
```

Reply:
While this is an interesting approach, I think all of this is just too much complexity for what its worth.

All I really want to handle here is injecting the has_leading_space_ parameter in a common place, and delegating to `AddLexedToken`.

I've re-done this code to not waste time on all the type checked versions. It turns out that for the vast majority, there was only one caller anyways so we were trading a bug here for a bug there which seems like not a win.